client/src/graphics/draw.rs

```diff
@@ -265,7 +265,7 @@ impl Draw {
     ) {
         let draw_started = Instant::now();
         let view = sim.view();
-        let projection = frustum.projection(0.01);
+        let projection = frustum.projection(0.005);
```

Comment:
Should this be derived from the radius, or even the default value, directly? A hard-coded constant is liable to become obsolete again.

Reply:
I'm not sure. One option is to make it extremely small. I believe `frustum.rs` uses a trick to take advantage of floating point to better distribute z-values in the depth buffer, which means that we should be able to go much smaller without z-fighting.

If we do want to make it configurable, it probably makes sense to put it in the client-specific config, and the near-plane distance can be given in "meters". However, there is an argument for putting it in sim_config so that the server can tune it based on other parameters that live in sim_config.

Reply:
It looks like the link that explains the reverse depth buffer in `frustum.rs` is broken. I can find it in the Wayback Machine, but an alternative link that describes the same thing is at https://developer.nvidia.com/content/depth-precision-visualized.

Even setting znear to 1.0e-30, I don't notice any z-fighting. It does break if I set it to 0, although there may be a workaround. Even if there's a workaround, it could potentially result in some pathological visual scenarios when going through a wall.

In short, for now, I would be in favor of just setting it to 1.0e-4 or something for now.

Reply:
I wouldn't read too much into visual tests using our current scene, since voxel geometry won't tend to have overlapping nearby faces very often.

> If we do want to make it configurable

I didn't mean to suggest that it be independently configurable, but rather that it be the largest value that the player collision geometry (governed by the server) ensures won't easily result in clipping.

> it looks like the the main fundamental limitation to decreasing znear is that typical depth values are scaled proportional to znear

Does that analysis account for the perspective divide?

Reply:
> I wouldn't read too much into visual tests using our current scene, since voxel geometry won't tend to have overlapping nearby faces very often.

That's a fair point. I've mainly gone off of mathematical analysis, and the visual tests were just final sanity checks. The resource at https://web.archive.org/web/20220807195222/http://dev.theomader.com/depth-precision/ has an interactive module that helps illustrate this benefit.

> I didn't mean to suggest that it be independently configurable, but rather that it be the largest value that the player collision geometry (governed by the server) ensures won't easily result in clipping.

My main concern there is that it causes configuration values to have second-order effects. Unless we have a clear reason that we want `znear` to be as large as possible without causing clipping, keeping it fixed leaves things more predictable.

> Does that analysis account for the perspective divide?

It does. With `zfar == 1`, we have `za == zb == -znear / (znear - 1.0)`. Looking at only the z and w coordinates, the projection matrix is
```
  za   zb
-1.0  0.0
```
If you multiply this by the column vector `[-z*w, w]`, where `w` is arbitrary, and `z` is between 0 and 1 (I negated z in `-z*w` because our projection matrix operates on vectors with negative z values), then you get the column vector
```
znear*w*(1-z) / (1-znear)
z*w
```
which divides out to `znear*(1-z) / (1-znear)`, which is roughly proportional to `znear` when `znear` is small.

Reply:
Good enough for me! Can always revisit if we see issues in practice.